forem/forem
# Create routes for i18n

## What type of PR is this? (check all applicable)

## Description
This PR is starting to lay the groundwork to support i18n and viewing Forem in various languages. This introduces the routing structure we want to use in the app by nesting all routes under an optional scope. This will allow routes such as `/lang-fr-ca/username`. As we build on i18n, this is how users will access Forem in various languages (such as French Canadian).

Setting this up as a "sub-folder" enables us to better approach SEO, caching, and server-side code in the future. We are doing it as `/lang-fr-ca/` as opposed to the more widely recognized `/fr/ca` approach to give ourselves more flexibility on how we want to handle future changes.

The loose gameplan here is to lay the groundwork (routes, caching, etc) so that the app has the ability to understand the language preference to build on top of later.

_Note: we'll want to continue to lean into using path helpers instead of hardcoded strings so that locale works across the app._

**TODO**
- [ ] Add validations to Organizations and pages to prevent the use of `lang-`.
- [ ] Add note about moving routes outside of locale
- [ ] Fix specs
- [ ] Add a few specs around locale

## Related Tickets & Documents
https://forem.dev/vaidehijoshi/what-internationalization-features-should-we-support-4kl

## QA Instructions, Screenshots, Recordings
1. Fire up the app locally.
2. Click around several routes to ensure they aren't broken.
3. Add `/lang-fr-ca/` to the front of some routes and see that it still works (i.e. http://localhost:3000/lang-fr-ca/dashboard).
4. Try to update your username to `lang-anything` and verify it shows an error.
5. Visit a user's profile to ensure the route still works.
6. Signup with some social auth to ensure the `omniauth_callback` routes are still working as expected.

That's it 🎉 

## Added tests?

## Added to documentation?
- [x] no documentation needed - I'm going to do a "state of i18n" writeup after this PR to explain the current state, where we're trying to go, and how the community can help.

![spinning_globe](https://media.giphy.com/media/CJBvaVsKM7tNC/giphy.gif)
